### ISSUE TITLE
cuda memory leak with requires_grad=True

### ISSUE THREAD

**Issue by i-cherkasov** [pytorch/pytorch#16303]
## 🐛 Bug

CUDA memory is not released when tensor is moved to CPU if _requires_grad=True_.

## To Reproduce

```python
import torch
 
n = int(1e7)
batch_sz = int(n / 10)

x = torch.tensor([1, 2], requires_grad=True, dtype=torch.float32, device="cpu")
# x = torch.tensor([1, 2], dtype=torch.float32, device="cpu")

# some huge tensor not fitting into the gpu memory
z = torch.randn(2, n, dtype=torch.float32, device="cpu")
 
batch_lst = []
 
for i in range(0, n, batch_sz):
    batch_lst.append(
        # do some heavy-weight computations on gpu
        torch.max((
            x.view([2, 1]).cuda() * \
            z[:, i:(i + batch_sz)].cuda()
            )**2,
            dim=0)[0].cpu() # and move back to cpu
    )
    print("allocated: %.2fG" % (torch.cuda.memory_allocated() / 1e9),
          "cached: %.2fG" % (torch.cuda.memory_cached() / 1e9))
 
# aggregate batches on cpu
f = torch.mean(torch.cat(batch_lst))
print("f:", f, f.device)
print("allocated: %.2fG" % (torch.cuda.memory_allocated() / 1e9),
      "cached: %.2fG" % (torch.cuda.memory_cached() / 1e9))
 
# gradient on cpu
g = torch.autograd.grad(f, x, retain_graph=True, create_graph=True)[0]
print("g:", g, g.device)
print("allocated: %.2fG" % (torch.cuda.memory_allocated() / 1e9),
      "cached: %.2fG" % (torch.cuda.memory_cached() / 1e9))

# 1st row of hessian on cpu
h1 = torch.autograd.grad(g[0], x, retain_graph=True, create_graph=True)[0]
print("h1:", h1, h1.device)
print("allocated: %.2fG" % (torch.cuda.memory_allocated() / 1e9),
      "cached: %.2fG" % (torch.cuda.memory_cached() / 1e9))

# 2nd row of hessian on cpu
h2 = torch.autograd.grad(g[1], x, retain_graph=True, create_graph=True)[0]
print("h2:", h2, h2.device)
print("allocated: %.2fG" % (torch.cuda.memory_allocated() / 1e9),
      "cached: %.2fG" % (torch.cuda.memory_cached() / 1e9))
```

Output:
```
allocated: 0.02G cached: 0.04G
allocated: 0.05G cached: 0.06G
allocated: 0.07G cached: 0.09G
allocated: 0.10G cached: 0.11G
allocated: 0.12G cached: 0.14G
allocated: 0.15G cached: 0.16G
allocated: 0.17G cached: 0.18G
allocated: 0.20G cached: 0.21G
allocated: 0.22G cached: 0.23G
allocated: 0.24G cached: 0.26G
f: tensor(4.3893, grad_fn=<MeanBackward1>) cpu
allocated: 0.24G cached: 0.26G
g: tensor([1.1008, 3.8390], grad_fn=<AddBackward0>) cpu
allocated: 0.49G cached: 0.51G
h1: tensor([1.1008, 0.0000], grad_fn=<AddBackward0>) cpu
allocated: 0.73G cached: 0.76G
h2: tensor([0.0000, 1.9195], grad_fn=<AddBackward0>) cpu
allocated: 0.98G cached: 1.00G
```

## Expected behavior

Smth. along these lines (comment out _requires_grad=True_ to reproduce, but then autograd will throw an error):

```
allocated: 0.00G cached: 0.03G
allocated: 0.00G cached: 0.03G
allocated: 0.00G cached: 0.03G
allocated: 0.00G cached: 0.03G
allocated: 0.00G cached: 0.03G
allocated: 0.00G cached: 0.03G
allocated: 0.00G cached: 0.03G
allocated: 0.00G cached: 0.03G
allocated: 0.00G cached: 0.03G
allocated: 0.00G cached: 0.03G
f: tensor(4.3877) cpu
allocated: 0.00G cached: 0.03G
```

Intuitively, I expected to see forward prop running on cuda, while back prop on cpu.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 1.0.0
 - OS (e.g., Linux): Ubuntu 18.04.1 LTS
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.6.7 
 - CUDA/cuDNN version: 10.0
 - GPU models and configuration:
 - Any other relevant information:

## Additional context

Same thing observed on a Windows box.

**Comment by soumith** [comment]
you are holding onto the graph inputs needed to compute gradients on the GPU.

More specifically, the `**2` here is for the operation `x^2`, and it's gradient is `2*x`. If you see, the input to `**2`, it's on the GPU (i.e. the output of `torch.max`. You have two options I think

1. put the whole `torch.max + **2` operation in a `with torch.no_grad():` block -- recommended and applies to any general operation.
2. move the output of `torch.max` to `.cpu()`, before you do `**2`
